File: Code/Swift/overloading.swift
protocol IReducible {
    init?<T: BinaryInteger>(exactly: T)
    static func * (lhs: Self, rhs: Self) -> Self
    static func * (lhs: Self, rhs: Int) -> Self
}

protocol IPrintable {
    func output()
}


struct MyType {
    var n: Int
}

extension MyType: IReducible, IPrintable {
    init?<T: BinaryInteger>(exactly: T) {
        self.n = Int(exactly)
    }
    static func * (lhs: MyType, rhs: MyType) -> MyType {        
        return MyType(n: lhs.n * rhs.n)
    }
    static func * (lhs: MyType, rhs: Int) -> MyType {        
        return MyType(n: lhs.n * rhs)
    }
    func output() {
        print("I am: \(self.n)") 
    }
}

extension Float64: IReducible, IPrintable {
    static func * (lhs: Float64, rhs: Int) -> Float64 {        
        return lhs * Float64(rhs)
    }
    func output() {
        print("I am: \(self)")
    }
}

func prod<T: IReducible>(arr: [T]) -> T {
    var res: T
    res = T(exactly: 1)!
    for i in 0 ..< arr.count {
        res = res * arr[i]
    }
    return res
}

func products<T: IReducible & IPrintable,
              R: IReducible & IPrintable>(at: inout [T], ar: inout [R]) {    
    var st: T
    var sr: R

    let ai: [Int] = [4,3,2,1]
    
    for i in 0 ..< at.count {
        at[i] = (at[i] * at[i]) * ai[i]
    }

    for i in 0 ..< ar.count {
        ar[i] = (ar[i] * ar[i]) * ai[i]
    }

    // calculate reduction of arrays of both types
    st = prod(arr: at)
    sr = prod(arr: ar)    
    
    st.output()
    sr.output()
}

func main() {

    var at: [MyType]  = [MyType(n:1),MyType(n:2),MyType(n:3),MyType(n:4)]
    var ar: [Float64] = [1.0,2.0,3.0,4.0]

    // print the results
    products(at: &at, ar: &ar)
}

main()

File: Code/Fortran/overloading.f90
module basic_interfaces

   implicit none
   private

   public :: IReducible, IPrintable

   abstract interface :: IReducible
      function init(n)
         integer, intent(in) :: n
      end function init
      elemental function operator(*)(lhs,rhs) result(res)
         type(itself), intent(in) :: lhs, rhs
         type(itself)             :: res
      end function operator(*)
      elemental function operator(*)(lhs,rhs) result(res)
         type(itself), intent(in) :: lhs
         integer,      intent(in) :: rhs
         type(itself)             :: res
      end function operator(*)
   end interface IReducible
   
   abstract interface :: IPrintable
      subroutine output()
      end subroutine output
   end interface IPrintable

end module basic_interfaces

module my_type

   use basic_interfaces, only: IReducible, IPrintable

   implicit none
   private
   
   type, public, sealed :: MyType
      private
      integer :: n
   end type MyType

   implements (IReducible,IPrintable) :: MyType
      initial :: init
      procedure, nopass :: operator(*) => multiply, multiply_by_int
      procedure, pass   :: output
   end implements MyType

contains
   
   function init(n) result(self)
      integer, intent(in) :: n
      type(MyType)        :: self
      self%n = n
   end function init
   
   elemental function multiply(lhs,rhs) result(res)
      type(MyType), intent(in) :: lhs, rhs
      type(MyType)             :: res
      res%n = lhs%n * rhs%n
   end function multiply

   elemental function multiply_by_int(lhs,rhs) result(res)
      type(MyType), intent(in) :: lhs
      integer,      intent(in) :: rhs
      type(MyType)             :: res
      res%n = lhs%n * rhs
   end function multiply_by_int
   
   subroutine output(self)
      type(MyType), intent(in) :: self
      write(*,'(a,i8)') "I am: ", self%n
   end subroutine output

end module my_type

module real_type

   use basic_interfaces, only: IReducible, IPrintable

   implicit none
   private
   
   implements (IReducible,IPrintable) :: real
      procedure, pass :: output
   end implements real

contains

   subroutine output(self)
      real, intent(in) :: self
      write(*,'(a,f8.5)') "I am: ", self
   end subroutine output

end module real_type

program main

   use basic_interfaces, only: IReducible, IPrintable
   use my_type,          only: MyType

   implicit none

   integer      :: i
   type(MyType) :: at(4)     ! Array of MyType 
   real         :: ar(4)     ! Array of real type
   
   ! initializations
   at = [(MyType(i),i=1,4)]  ! Use of user-defined initializer/constructor
   ar = [(real(i),i=1,4)]    ! Use of built-in initializer (i.e. cast)
   
   call products(at,ar)
   
contains

   subroutine products{IReducible,IPrintable :: T,R}(at,ar)
      type(T), intent(in) :: at(4)
      type(R), intent(in) :: ar(4)
      type(T) :: st                      ! Scalar of type(T)
      type(R) :: sr                      ! Scalar of type(R)
      st = prod( (at * at) * [4,3,2,1] ) ! Reduce arrays to scalars after
      sr = prod( (ar * ar) * [4,3,2,1] ) ! elementwise multiplication operations
      call st%output()                   ! Print the results
      call sr%output()                   ! (13824 in both cases)
   end subroutine products

   function prod{IReducible :: T}(arr) result(res)
      type(T), intent(in) :: arr(:)
      type(T)             :: res
      integer :: i
      res = T(1)                         ! Use initializer for cast
      do i = 1, size(arr)
         res = res * arr(i)              ! Use operator(*) for reduction
      end do
   end function prod
   
end program main
